interp/src/stk_env.rs

```diff
@@ -0,0 +1,896 @@
+    /// a.set("hi!", 1);
+    /// a.set("bye", 0);
+    /// a.new_scope();
+    /// a.set("hi!", 2);
+    /// assert_eq!(*a.get(&"hi!").unwrap(), 2);
+    /// assert_eq!(*a.get(&"bye").unwrap(), 0);
+    /// let a = a.smoosh_once();
+    /// assert_eq!(*a.get(&"hi!").unwrap(), 2);
+    /// assert_eq!(*a.get(&"bye").unwrap(), 0);
+    ///```
+    ///the following should panic:
+    ///```should_panic
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hi!", 1);
+    /// a.set("bye", 0);
+    /// let a = a.smoosh_once(); //not enough scopes to smoosh
+    ///```
+    pub fn smoosh_once(self) -> Self {
+        //move head to a sep variable
+        let wr_head = self.head;
+        //now move the head of the tail into the head of the smoosher
+        let interm_tail = self.tail;
+        let (new_head, new_tail) = interm_tail.split();
+        if let Some(mut new_head) = new_head {
+            for (k, v) in wr_head {
+                new_head.insert(k, v);
+            }
+            Smoosher {
+                head: new_head,
+                tail: new_tail,
+            }
+        } else {
+            panic!("Could not smoosh, because [self] has less than two scopes")
+        }
+    }
+
+    /// ```text
+    /// Consumes [self] and returns a Smoosher in which
+    /// all bindings found in the top [levels] scopes of [self] to the [levels]th
+    /// scope are merged. [smoosh(0)] has no effect, while [smoosh(1)] is equivalent to
+    /// [smoosh_once].
+    /// ```
+    /// # Guarantees
+    ///  If [self] has **n** scopes, then
+    /// [self.get(K)] == [self.smoosh(n).get(K)] for all K bound in [self]
+    ///
+    ///  # Invariant
+    /// ```text
+    /// None of the top [levels] scopes may be the root of any fork
+    /// ```
+    /// # Panics
+    /// ```text
+    /// Panics if [levels] >= # of scopes in the Smoosher, or any of the top
+    /// [levels] scopes are the roots of any existing forks.
+    /// ```
+    /// # Examples
+    /// ## Pictorial Examples
+    /// ```text
+    ///  [A]  
+    ///   |
+    ///  [B]
+    ///   |     => [A.smoosh(2)] => [A U B\A U C\(A U B\A)]
+    ///  [C]
+    /// ```
+    /// ## Code Examples
+    ///```
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hi!", 1);
+    /// a.new_scope();
+    /// a.set("bye", 0);
+    /// a.new_scope();
+    /// a.set("hi!", 2);
+    /// let a = a.smoosh(2);
+    /// assert_eq!(*a.get(&"hi!").unwrap(), 2);
+    /// assert_eq!(*a.get(&"bye").unwrap(), 0);
+    ///```
+    ///the following should panic:
+    ///```should_panic
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hi!", 1);
+    /// let b = a.fork();
+    /// a.set("bye", 0);
+    /// let a = a.smoosh(1); //cannot smoosh; a fork point exists in a's tail
+    /// b.get(&"hi!");
+    ///```
+    pub fn smoosh(self, levels: u64) -> Self {
+        let mut tr = self;
```

Comment:
you could pass `mut self` rather than `self` and avoid needing to move into a new variable

Reply:
Maybe we will begin building the environment and seeing if that's more stylistic? I think it makes sense to keep `smoosh` and `merge` both functional b/c they are related, and `merge` must be functional.

Reply:
I'm not suggesting changing whether or not it takes ownership.

I mean literally `mut self` instead of `self`, not `&mut self`

Reply:
But again, this is just to remove variables that aren't specifically necessary